### PR TITLE
GSoC 2018: Setup Improvements Continue(libreehr)

### PR DESCRIPTION
**Pull request by muarachmann**
Rewriting the scripts lost some files in the process of rebasing , deleting and moving repos thank God i can find my way out. 
A cleaner branch and rapid commits like never before on their way

### CODE REVIEW

**Comment by muarachmann** [comment]
Continuation of the closed PR #1135 here

**Comment by muarachmann** [comment]
@tmccormi , @pri2si17-1997 , @aethelwulffe got a cleaner branch and made a PR

**Comment by muarachmann** [comment]
screenshot from 2018-06-27 11-23-16
MySQL configuration page with ajaxing behind the scene

**Comment by muarachmann** [comment]
screenshot from 2018-06-27 11-28-26

Ajax response gotten for successfull login parameters incase of no cloning

**Comment by aethelwulffe** [comment]
Ooooooooooh!  Aaaaaaaaaawe!  Aaaaaaaah!
LibreEHR AND Arduino!  We can hook up a massage chair motor controller and have it run while installing!  Or at least some cool LEDs!

**Comment by muarachmann** [comment]
@aethelwulffe some label permissions will do here

**Comment by muarachmann** [comment]
I have this error -> when i choose the option clone database `unable to open file tmp/setup_dumb.sql` i dont even see this file in my repo if i could be cleared up on what this does then i can solve this issue.

**Comment by muarachmann** [comment]
Then have this other problem when i dont click on clone db but select another site id it hooks up at the phase of inserting version with error `Error- setup failed: ERROR. Unable insert version information into database(#)`

**Comment by muarachmann** [comment]
screenshot from 2018-06-29 21-42-36

**Comment by muarachmann** [comment]
Why does this occurs and when does this occurs so i can kno what to do. I guess its probably a directory structure stuff. Here is my methodology since i had to create a separate module which is found in modules/setup, files directory had to change it the path in the installer(Databse class) since i had to move abt two to three directories above to get the base dir and stuffs. so i guess some places are getting confused with the path. I managed to change most of the path bu am sure deep deep something  is happening for it not to open that file. As for the second error i have no idea why it cant add version

I changed lines line 

`GLOBALS['OE_SITES_BASE'] = dirname(__FILE__) . '/../../../sites';`
    `$this->conffile  =  $GLOBALS['OE_SITE_DIR'] . '/sqlconf2.php';` a holder for now and works fine


I guess it all resolved around here if you could get me know what to do in this function i would gladly appreciate is it to point to a particular file or?? 

`private function get_backup_filename() {
        if (stristr(PHP_OS, 'WIN')) {
            $backup_file = 'C:/windows/temp/setup_dump.sql';
        }
        else {
            $backup_file = '/tmp/setup_dump.sql';
        }
        return $backup_file;
    }`

**Comment by muarachmann** [comment]
Sucess screen can be seen below
screenshot from 2018-06-29 22-19-37
screenshot from 2018-06-29 22-19-42

screenshot from 2018-06-29 22-19-46

**Comment by muarachmann** [comment]
@pri2si17-1997  , @tmccormi  , @aethelwulffe  :point_up:

**Comment by tmccormi** [comment]
Likely permissions, that file is created to dump the src and reimport

**Comment by muarachmann** [comment]
Fixed the bug for initialization of user screen looks thus now. Will proceed to configuration of (php-GACL). while awainting an explation for the `$backup_file = '/tmp/setup_dump.sql';` cause i dont see it on my sys. though still figuring out the stuff :nerd_face: 

screenshot from 2018-06-30 01-40-02

**Comment by muarachmann** [comment]
@tmccormi have you tested this PR? not to worry i have pointed the sqlconfig to sqlconfig2 so it wont disturb my current configuration

**Comment by muarachmann** [comment]
I doubted seeing a /tmp/....  i guess it is temporal. @tmccormi is this file created and deleted immediately after being used? or kept within the repo?

**Comment by tmccormi** [comment]
It's not a repo file. Just created, used and deleted.  Read the current setup code for details

**Comment by tmccormi** [comment]
I will test this weekend.

**Comment by muarachmann** [comment]
I doubted as much i guess(knew its something temp). Okay going keenly through and also checking my file permission cause i think i did a *chown -R www-data repo* lets fetch this out

**Comment by muarachmann** [comment]
@tmccormi , @aethelwulffe after the creation of database and initialization of user i have *158* rows returned on a *show tables* for my newly created db. Please assert if this is ok

**Comment by tmccormi** [comment]
Compare what you get using the standard (current) setup process and see if it's the same, you don't need us to do that for you.

**Comment by muarachmann** [comment]
screenshot from 2018-07-07 19-48-48

Running of the setup.sh script to install packages

**Comment by muarachmann** [comment]
screenshot from 2018-07-07 19-49-01
screenshot from 2018-07-07 19-49-05

After accepting the install procedure

**Comment by muarachmann** [comment]
screenshot from 2018-07-07 19-49-07

Successful install and opens the browser to the step left off

**Comment by aethelwulffe** [comment]
Nice.  Finally getting to testing this.  I assume more coming pretty soon here?

**Comment by muarachmann** [comment]
yeah @aethelwulffe more commits coming soon and final screen will be uploaded

**Comment by muarachmann** [comment]
will edit the index file at the end

**Comment by pri2si17-1997** [comment]
@muarachmann 

I tested this PR(till latest commit) and I must say, you have done great job. :+1: 

Also I have some queries and suggestions, which I thought of. I will mention it below one by one : 

**Suggestions**

- 
scrolling_suggestion

I would say, that you create a scrollbar inside your area, instead of using browser scrollbar. That can be improved later on. Not to worry.

- 
one_progress_suggestion

Here I think that there should be one progress notification only. What you are doing is great, but It do not make much sense to have same message twice. Means one time you are popping into progress bar, and at the same time, you have toast with same message above. IMO you should remove toast message and only keep progress bar. What you can do is, make a tooltip on hover to progress bar, if it contain  long message. But it looks good.

- 
edit_theme_screenshot

This one is from theme section, where you have radio buttons to select default theme. In screenshot, just crop it and keep only main portion. i.e., remove tabs and footer. 

- 
no_way_to_get_back

Here I can't find any way to get back to previous screen, if any error occurs. That should be there.

- 
error_if_database_fails

Now consider this one. What if due to any reason it failed. You should have something to start the process all over again. I tried to edit the info, pressed continue button, but it didn't worked. You can see on browser tab that it is loading, but ideally it should stop. Progress bar should disappear and you should have a screen with error message, with link to start the process all over again.

- 
no_continue

There is no continue or next button. Is it by design? If yes then any specific reason? or there should be link to login page of application.

These are my findings. I would like @teryhill @tmccormi @aethelwulffe to have their opinion on it. Also, these steps will happen from setup page (by url : `localhost/lh-ehr`), not from  (`http://localhost/ranchman/lh-ehr/modules/setup/start_up.php`). Am I correct?

**Comment by muarachmann** [comment]
@pri2si17-1997 thanks very much for these corrections/modifications. I haven't yet finished hope to be in some days. I will considering removing the alert box i just wanted to display a feedback to the user. Also that is not the final for that step. will make a loader that will cover the screen and prevent user from entering further in the form. Then on an error will reset the form. For the theme preview it was just a demo will reshape the images. And that is the last screen before a congratulation page of the sort will appear asking the user to login. More over I havent worked on the error part handling that right away.

**Comment by muarachmann** [comment]
The last screen is the parameters.php screen that will have a skip button cause it is not really recommended. If a user skips it it will insert the stuffs into the table by default  and vice versa.

File: modules/setup/libs/js/setup.js [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+    // function to clear the timer if process has completed to a hundred percent
+    function completed() {
+        $("#ajaxAlert").removeClass("hidden alert-danger").addClass("alert-success");
+        $("#ajaxResponse").html("Completed");
+        $("#submitStep4").removeAttr("disabled").css("cursor","pointer");
+        $("#backStep4").removeAttr("disabled").removeClass("btnDisabled").css("cursor","pointer");
+        window.clearInterval(timer);
+        $("#databaseForm").off('submit').submit();
+    }
+
```
Comment (muarachmann) [inline modules/setup/libs/js/setup.js]: Form dont seem to reset at this level

Comment (muarachmann) [reply modules/setup/libs/js/setup.js]: @pri2si17-1997 please try and see if you can help me out thanks

File: modules/setup/libs/js/setup.js [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+    // function to clear the timer if process has completed to a hundred percent
+    function errorComplete(msg) {
+        $("#ajaxAlert").removeClass("hidden alert-success").addClass("alert-danger");
+        $("#ajaxResponse").html("Error- setup failed: " + msg);
+        $("#submitStep4").removeAttr("disabled").css("cursor","pointer");
+        $("#backStep4").removeAttr("disabled").removeClass("btnDisabled").css("cursor","pointer");
+        window.clearInterval(timer);
+        $("#databaseForm").off('submit').reset();
+
+    }
```
Comment (muarachmann) [inline modules/setup/libs/js/setup.js]: After calling error complete

**Comment by aethelwulffe** [comment]
Hi Mua.  Moving along here nicely.  I think testing this will require my brains to be working, and to proceed with notes and the like.  I will run tests in the morning with an eye towards multi-site installations.

**Comment by muarachmann** [comment]
@aethelwulffe everything works fine just the error handling parts. Assuming an error occurs on setup procedure lets say database already exist.

**Comment by tmccormi** [comment]
I've downloaded this for testing in a clean VM during my travels this week...

**Comment by muarachmann** [comment]
@aethelwulffe, @tmccormi , @pri2si17-1997  i am done with the setup now i have this problem to redirect to the appropriate site. For noe the module detects if you have already libreehr installed. What i want is more than that i want it to tell me how many installations. That is it should read each sqlconf file in the sites directory and return the directory where `config = 1`. This would be helpful so that if we detect a site we prompt the user to enter the site of his/her choice :wink: . You could test the module it works ok that is creates database , clone , wite basic configurations etc. Now if you want to login into a specific site you will have to do so manually by adding ?site=default4 for example. I wish to remove this feature and replace with what i have mentioned above any ideas would be welcomed thanks

**Comment by aethelwulffe** [comment]
OK, in PHP, iteratoriterator() with a mask should be able to give you a listing of sites that are configured in a given instance.  /default/ may be live, but should not be for a new install.  

A feature to allow this would be on the landing page itself.  Set the login page to refresh(or that client side stuff) after it builds a list of site directories and has the user select one.  THEN it should be able to make a DB connection and still check user credentials.  You could make the landing page/site selection totally separate as a first step if desired.  Honestly, there would be nothing wrong with popping a minimal content management system (Maybe Naveens "Crow CMS") in front of everything, and having a link builder in there, as well as allowing a clinic to simply use that for their basic website.  Links to the patient portal could function the same way.

**Comment by aethelwulffe** [comment]
I agree that we need the whole SITES thing figured out before deploying this, but you don't have to make it too complex for now.  Just something that generates a drop-down list of the available site directories to choose from would be good enough.

**Comment by muarachmann** [comment]
@aethelwulffe , @tmccormi , @teryhill i messed up my master till the point of re-fork. Will be closing this and opening a new one. Thank God tony had this squashed. i had some latest work but i can replicate that. Will push once everything is done. How far with the translations...

**Comment by muarachmann** [comment]
> I agree that we need the whole SITES thing figured out before deploying this, but you don't have to make it too complex for now. Just something that generates a drop-down list of the available site directories to choose from would be good enough. No need to create an object array to iterate through, right? It is just a menu builder.
> `array scandir ( string $directory [, int $sorting_order = SCANDIR_SORT_ASCENDING [, resource $context ]] )`

This is helpful a lot

**Pull request closed without merging**